datreant/backends/statefiles.py

```diff
@@ -0,0 +1,448 @@
+"""
+Interface classes for state files.
+
+"""
+
+import os
+import warnings
+import json
+
+from six import string_types
+
+import datreant
+from .core import FileSerial
+
+
+class MixinJSON(object):
+    def _deserialize(self, handle):
+        return json.load(handle)
+
+    def _serialize(self, record, handle):
+        json.dump(record, handle)
+
+
+class TreantFile(MixinJSON, FileSerial):
+    def __init__(self, filename, logger=None, **kwargs):
+        """Initialize Treant state file.
+
+        This is the base class for all Treant state files. It generates data
+        structure elements common to all Treants. It also implements
+        low-level I/O functionality.
+
+        :Arguments:
+            *filename*
+                path to file
+            *logger*
+                Treant's logger instance
+
+        :Keywords:
+            *treanttype*
+                Treant type
+            *name*
+                user-given name of Treant object
+            *categories*
+                user-given dictionary with custom keys and values; used to
+                give distinguishing characteristics to object for search
+            *tags*
+                user-given list with custom elements; used to give
+                distinguishing characteristics to object for search
+            *version*
+                version of datreant file was generated with
+
+        .. Note:: kwargs passed to :meth:`create`
+
+        """
+        super(FileSerial, self).__init__(filename, logger=logger)
+
+        # if file does not exist, it is created; if it does exist, it is
+        # updated
+        try:
+            self.create(**kwargs)
+        except OSError:
+            # in case the file exists but is read-only; we can't update but may
+            # still want to use it
+            if os.path.exists(self.filename):
+                pass
+            # if the file doesn't exist, we still want an exception
+            else:
+                raise
+
+    def _init_record(self):
```

Comment:
It's used to initialize the data structure in case it doesn't exist. When a `TreantFile` object is created, it first creates its proxy file for applying locks, and then calls its `create` method, which will then call methods like `update_schema`, `update_version`, `add_tags`, and `add_categories`. However, these all require a write to the state file, but what if the file doesn't actually exist (such as when this is a new `Treant`)? Pulling the current version of the data structure will fail when each of these methods is called.

A smooth way to make this work is to have the `_write` decorator handle it. A lock gets applied to the proxy file, but if the statefile itself doesn't exist yet, the data structure it would have is initialized, and this is later written out when the method that was wrapped finishes.

See [here](https://github.com/datreant/datreant/blob/exp-dynamic_data/datreant/backends/core.py#L401) for where it lives in the `_write` decorator. I think there are other ways of handling the init issue for the data structure that lives in the state file, but I think this one has nice features.